torchdata/datapipes/iter/util/randomsplitter.py

```diff
@@ -0,0 +1,167 @@
+            else:
+                raise KeyError(f"`target={target}` does not match any key in `weights`.")
+
+
+class _RandomSplitterIterDataPipe(IterDataPipe):
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe,
+        total_length: int,
+        weights: Dict[T, Union[int, float]],
+        seed,
+    ):
+        self.source_datapipe: IterDataPipe = source_datapipe
+        self.total_length = total_length
+        self._seed = seed
+        self.norm_weights = self.normalize_weights(weights, total_length)
+        self.keys = list(weights.keys())
+        self.key_to_index = {k: i for i, k in enumerate(self.keys)}
+        self.weights = [self.norm_weights[k] for k in self.keys]
+        self._rng = random.Random(self._seed)
+
+    def draw(self):
+        selected_key = self._rng.choices(self.keys, self.weights)[0]
+        self.weights[self.key_to_index[selected_key]] -= 1
+        return selected_key
+
+    @staticmethod
+    def normalize_weights(weights, total_length: int):
+        total_weight = sum(weights.values())
+        return {k: round(float(w) * total_length / total_weight) for k, w in weights.items()}
+
+    def reset(self):
+        self._rng = random.Random(self._seed)
+        self.weights = [self.norm_weights[k] for k in self.keys]
+
+    def set_seed(self, seed):
+        """
+        Update the `seed`. The new `seed` will be used in the next iteration.
+        """
+        self._seed = seed
+
+    def __getstate__(self):
+        if IterDataPipe.getstate_hook is not None:
+            return IterDataPipe.getstate_hook(self)
+        state = (
+            self.source_datapipe,
+            self.total_length,
+            self._seed,
+            self.norm_weights,
+            self.keys,
+            self.key_to_index,
+            self.weights,
+            self._rng.getstate(),
+        )
+        return state
+
+    def __setstate__(self, state):
+        (
+            self.source_datapipe,
+            self.total_length,
+            self._seed,
+            self.norm_weights,
+            self.keys,
+            self.key_to_index,
+            self.weights,
+            rng_state,
+        ) = state
+        self._rng = random.Random()
+        self._rng.setstate(rng_state)
+
+
+class SplitterIterator(IterDataPipe):
+    def __init__(self, main_datapipe: _RandomSplitterIterDataPipe, target: T):
+        self.main_datapipe = main_datapipe
+        self.target = target
+
+    def __iter__(self):
+        self.main_datapipe.reset()
+        for sample in self.main_datapipe.source_datapipe:
+            if self.main_datapipe.draw() == self.target:
+                yield sample
+
+    def __len__(self):
+        return self.main_datapipe.norm_weights[self.target]
+
+    def set_seed(self, seed):
+        """
+        Update the `seed`. The new `seed` will be used in the next iteration.
+        """
+        self.main_datapipe.set_seed(seed)
```

Comment:
I was talking about the `set_seed` here. The way I think `apply_seed` should work is:
- Traverse the graph and list all datapipes
- Find the datapipes with `set_seed` api among the list of them
- Set seeds automatically by `DataLoader2` + `ReadingService`

If we keep `set_seed` to both parent and children, it would do seeding `num_of_children + 1` times. It's kind wasting time even thought the seed is still aligned across them

Reply:
In terms of behavior, IIUC that would change the results of `random_split` for every epoch right? I would guess users prefer not to do that? Do you agree?

So we should either remove or rename `set_seed` to avoid `apply_seed` from using it.

Reply:
Let's fold this discussion to the other comment thread.